app-backend/db/src/main/scala/TaskSessionDao.scala

```diff
@@ -334,8 +334,8 @@ object TaskSessionDao extends Dao[TaskSession] {
   def filterLabel(
       sessionId: UUID,
       labelId: UUID
-  ): Dao.QueryBuilder[AnnotationLabelWithClasses] =
-    AnnotationLabelDao.query
+  ): Dao.GroupQueryBuilder[AnnotationLabelWithClasses] =
+    AnnotationLabelDao.withClassesQB
```

Comment:
I just want some clarity on this, for my edification

is the change above replacing this:
```
  val selectF: Fragment = fr"SELECT" ++
    selectFieldsF ++ fr", classes.class_ids as annotation_label_classes FROM " ++
    Fragment.const(tableName) ++
    fr""" JOIN (
      SELECT annotation_label_id, array_agg(annotation_class_id) as class_ids
      FROM """ ++ Fragment.const(joinTableName) ++ fr"""
      GROUP BY annotation_label_id
    ) as classes ON """ ++ Fragment.const(tableName) ++ fr".id = " ++
    fr"classes.annotation_label_id"
```

for this:
```
  val withClassesQB: Dao.GroupQueryBuilder[AnnotationLabelWithClasses] =
    Dao.GroupQueryBuilder[AnnotationLabelWithClasses](
      fr"SELECT" ++ selectFieldsF ++ fr", array_agg(annotation_class_id) as annotation_label_classes",
      fr"annotation_labels JOIN annotation_labels_annotation_label_classes ON annotation_labels.id = annotation_labels_annotation_label_classes.annotation_label_id",
      NonEmptyList.of(fr"id"),
      Nil
    )
```

how is it that these fundamentally differ? there is not a full join on both of them?

Reply:
The important part is:

```
JOIN (
      SELECT annotation_label_id, array_agg(annotation_class_id) as class_ids
      FROM """ ++ Fragment.const(joinTableName) ++ fr"""
      GROUP BY annotation_label_id
    )
```

postgres can't push filters into subqueries, so the `JOIN` evaluates that groupby on the whole table before it gets joined and filtered with the labels. The reason  we did that before is that the default `QueryBuilder` doesn't know how to use `group by` operations, so filters got inserted out of order with the group by statement if you tried to set it up, and postgres told you your syntax was wrong.

The `GroupQueryBuilder` understands that filters go before `group by` statements, so your `tableF` can be an arbitrarily complex join and the filters will be applied prior to that join (provided you don't have any subqueries in that arbitrarily complex join). https://github.com/raster-foundry/raster-foundry/blob/develop/app-backend/db/src/main/scala/Dao.scala#L151-L155

Reply:
gotcha, ok that makes sense